src/main/flight/imu.c

```diff
@@ -366,80 +380,36 @@ STATIC_UNIT_TESTED void imuUpdateEulerAngles(void)
-//   - wait for a 250ms period of low gyro activity to ensure the craft is not moving
-//   - use a large dcmKpGain value for 500ms to allow the attitude estimate to quickly converge
-//   - reset the gain back to the standard setting
-static float imuCalcKpGain(timeUs_t currentTimeUs, bool useAcc, float *gyroAverage)
+// Calculate gain based on
+// * how close the magnitude of the acc is to 1g
+// * how close the rotational rate is to 0
+// * how much we trust the current estimate
+static float imuCalcKpGain(float *deviation, const float baseKp, const float dt, const float accTrust)
 {
-    static enum {
-        stArmed,
-        stRestart,
-        stQuiet,
-        stReset,
-        stDisarmed
-    } arState = stDisarmed;
-
-    static timeUs_t stateTimeout;
-
-    const bool armState = ARMING_FLAG(ARMED);
-
-    if (!armState) {
-        // If gyro activity exceeds the threshold then restart the quiet period.
-        // Also, if the attitude reset has been complete and there is subsequent gyro activity then
-        //  start the reset cycle again. This addresses the case where the pilot rights the craft after a crash.
-        if (   (fabsf(gyroAverage[X]) > ATTITUDE_RESET_GYRO_LIMIT)  // gyro axis limit exceeded
-            || (fabsf(gyroAverage[Y]) > ATTITUDE_RESET_GYRO_LIMIT)
-            || (fabsf(gyroAverage[Z]) > ATTITUDE_RESET_GYRO_LIMIT)
-            || !useAcc                                              // acc reading out of range
-            ) {
-            arState = stRestart;
-        }
-
-        switch (arState) {
-        default: // should not happen, safeguard only
-        case stArmed:
-        case stRestart:
-            stateTimeout = currentTimeUs + ATTITUDE_RESET_QUIET_TIME;
-            arState = stQuiet;
-            // fallthrough
-        case stQuiet:
-            if (cmpTimeUs(currentTimeUs, stateTimeout) >= 0) {
-                stateTimeout = currentTimeUs + ATTITUDE_RESET_ACTIVE_TIME;
-                arState = stReset;
-            }
-            // low gain during quiet phase
-            return imuRuntimeConfig.dcm_kp;
-        case stReset:
-            if (cmpTimeUs(currentTimeUs, stateTimeout) >= 0) {
-                arState = stDisarmed;
-            }
-            // high gain after quiet period
-            return ATTITUDE_RESET_KP_GAIN;
-        case stDisarmed:
-            // Scale the kP to generally converge faster when disarmed.
-            return imuRuntimeConfig.dcm_kp * 10.0f;
-        }
-    } else {
-        arState = stArmed;
-        return imuRuntimeConfig.dcm_kp;
-    }
+    const float deviationGainMax = 100.0f * baseKp;
+    const float deviationGainMin = 0.1f * baseKp;
+    const float kpGain = scaleRangef(*deviation, 0.0f, 180.0f, deviationGainMin, deviationGainMax) * accTrust;
+
+    // decrease the estimated deviation based on how much we currently trust the acc
+    static const float heuristic = 0.1f;
+    const float omega = 2.0f * M_PIf * dt * (heuristic * kpGain);
```

Comment:
The short rationale is that since the filter should converge exponentially the trust in the current estimate should also converge exponentially. 
It's intended to use basically the same math as the PT1 filter, but I was quite tired when I wrote that code.
I've also changed the code to use the base kp gain now, since I had problem with the deviation converging to quickly during flight tests.
I will definitely revisit this part.